core: add TemporaryFileSystem= setting and 'tmpfs' option to ProtectHome=

yuwata opened the pull request:
This introduces a new setting `TemporaryFileSystem=`. The unit process can access the specified directories, but cannot access the files stored in the directories.
Also, this adds a new option `tmpfs` to `ProtectHome=`, which is equivalent to `TemporaryFileSystem=/root /run/user /home`.

This is an important preparation to support #7895, #7153 and #2780.

vcaputo commented on src/core/execute.c:
```diff
@@ -2295,6 +2294,12 @@ static int compile_bind_mounts(
 
         assert(h == n);
 
+        if (!strv_isempty(empty_directories) && !strv_isempty(context->empty_directories)) {
```
I admit I haven't spent much time looking at this code, but at a casual glance shouldn't context->empty_directories still be applied when empty_directories is empty at this point?  It looks like a non-empty context->empty_directories will just be ignored by this function if empty_directories is empty.  Is that intentional?

poettering replied:
yeah, i think @vcaputo is right

vcaputo commented on src/core/execute.c:
```diff
@@ -2369,7 +2374,7 @@ static int apply_mount_namespace(
                             needs_sandboxing ? context->read_only_paths : NULL,
                             needs_sandboxing ? context->inaccessible_paths : NULL,
-                            empty_directories,
+                            empty_directories ? : context->empty_directories,
```
Oh, is this why compile_bind_mounts() doesn't return empty_directories when context->empty_directories is the only one?

That's kind of ugly IMHO.

yuwata replied:
It might be. But this logic can reduce creating unnecessary copy of `context->empty_directories`.

vcaputo commented on test/test-execute/exec-emptydirectories.service:
```diff
@@ -0,0 +1,22 @@
+[Unit]
+Description=Test for BindPaths= and BindReadOnlyPaths=
```
Unimportant nit, but I think this is stale from whatever was used as a template in creating this...

yuwata replied:
Oh, thanks. I will update.

yuwata replied:
Updated.

vcaputo commented:
I gave a casual look over the code, made some comments, but it generally lgtm.

yuwata commented:
Also added `ProtectHome=empty`, which is equivalent to `EmptyDirectories=/root /run/user /home`.

poettering commented:
Hmm, what's the difference and benefit over InaccessiblePaths=?

GMLudo commented:
Hi @poettering,

To my understanding, it's impossible to do that with **InaccessiblePaths=**: It works for read-only access limitations on the file system, but not for a complete black-out, like you have in a chroot, where you can see only specific folders.
Another example : https://unix.stackexchange.com/questions/305586/how-to-whitelist-directories-for-units-in-systemd/387051#387051
As you can read in the answer: "According to this quote, it doesn't work".

The main interest: With this option, it will be possible to build easily a new complete view on the filesystem without the need to maintain a separate chroot.

For the sysadmins, they can apply the security upgrades on the "host", all daemons will have an updated view of their part of root, not need to upgrade each container, with more or less the same security level of a "complete" container with a separate root.

We use intensively systemd features to reduce our number of VMs and containers.
Our final goal is to enable the maximum of hardening options available in systemd to increase the security of daemons we are using on production, to try to reduce the risk of 0-day attack.

I have started this project with bind, I'm discussing on their mailing-list about this: https://lists.isc.org/pipermail/bind-users/2018-January/099437.html

Thanks for your remarks.

topimiettinen commented:
I think #2780 is also related.

topimiettinen commented:
What about DAC permissions, ACLs, SELinux labels, mount flags (noexec/nodev/nosuid/ro/rw) of the original directory? Shouldn't they also be reapplied to the empty directory and/or let the admin configure them in the service file?

sourcejedi commented:
Documentation here doesn't mention the permissions of the created tmpfs, i.e. who can write to it.  I'm not saying it's wrong to set it to 0755, but that's not the default permission for a tmpfs, so I think the docs should be more explicit somehow.

yuwata commented:
Updated. Mainly adds more documentation.

@poettering 
Example,
```
EmptyDirectories=/etc
ReadOnlyPaths=/etc/systemd
```
This hides any configuration under /etc except for /etc/systemd. Several RFEs request such a function, and we already internally use it for '/private' `StateDirectory=`. This just expose it as a new setting.

@topimiettinen Sorry, I do not have enough knowledge about them. What kind of information should be inherit from the original directory?

@sourcejedi Initially they are mounted with mode=755, but after that they are remounted as read-only.
I've updated the documents to mention that.

topimiettinen commented:
@yuwata I think safest would be all of them, so that the result is not different from taking the original directory and removing its contents.

yuwata commented:
@topimiettinen Do you concern them on just EmptyDirectories= setting? Or on BindPaths= ?

poettering commented:
I can see the usecase, but I am not sure the name "EmptyDirectory=" is really that great for this, after all for it to be useful (i.e. to be more than InaccessibleDirectory=) it will *not* be empty. I mean, the fact that it doesn't have to be entirely empty, and doesn't cancel out any mounts below it is kinda the key feature of it, no?

In "systemd-nspawn" we already have the `--tmpfs=` option (which is exposes as TemporaryFileSystem= in .nspawn files), which pretty much is what you are trying to do here too, right? Maybe we should just reuse that name for the same concept here? i.e. introduce this as TemporaryFileSystem= option in unit files?

poettering commented on TRANSIENT-SETTINGS.md:
```diff
@@ -168,6 +168,7 @@ All execution-related settings are available for transient units.
 ✓ ReadWritePaths=
 ✓ ReadOnlyPaths=
 ✓ InaccessiblePaths=
+✓ EmptyDirectories=
```
the commit's title is wrong btw, it says "ExecDirectories=", but should say `EmptyDirectories=`

yuwata commented:
@poettering Yeah, sounds good. I will rename them.

poettering commented:
(also I have the suspicion this should take an optional mount option string too, the same way as nspawn's --tmpfs= takes one, separated by a colon)

poettering commented on src/core/execute.c:
```diff
@@ -2250,7 +2251,7 @@ static int compile_bind_mounts(
                          * tmpfs that makes it accessible and is empty except for the submounts we do this for. */
 
-                        private_root = strjoin(params->prefix[t], "/private");
+                        private_root = strjoin("+", params->prefix[t], "/private");
```
hmm, why this?

GMLudo commented:
Hi @yuwata, I permitted me to send you a gentle ping in case of you have some time to answer on review questions :-)
I'm interested by this feature, but I don't have the technical level to contribute on it :-(
However, I can do some tests with it.

Have a nice week.

yuwata commented:
@GMLudo I was tackling other issues recently. I guess that I can update this not so far future, I hope this weekend :-) Testing, commenting, etc., any feedbacks are of course welcome. Thank you.

GMLudo commented:
@yuwata Pretty cool, thanks for your help :-)
At the same time, I'm reading how to compile and install systemd from sources to be able to test this PR, because for now I'm using Debian backports.
To my understanding, the easiest way is systemd-nspawn: https://www.freedesktop.org/wiki/Software/systemd/VirtualizedTesting/
I will try to propose extra documentation for the wiki to do that.

yuwata commented:
Updated. Now the name of new setting is changed to `TemporaryFileSystem=`. As commented by @poettering, now each mount point can be suffixed with a colon and mount options.

Please take a look.

sourcejedi commented on src/core/namespace.c:
```diff
@@ -724,6 +793,20 @@ static int mount_empty_dir(const MountEntry *m) {
+        (void) mkdir_p_label(mount_entry_path(m), 0755);
+        (void) umount_recursive(mount_entry_path(m), 0);
+
+        if (mount("tmpfs", mount_entry_path(m), "tmpfs", MS_NODEV|MS_STRICTATIME, mount_entry_option(m)) < 0)
```
Hmm.

Will the kernel help us by converting "noatime" into MS_NOATIME and override the constant we passed?

Similar question with "ro", "nosuid" etc.

If not, maybe libmount counld help us.

sourcejedi replied:
It *might* not be good to rely on the raw kernel behaviour. Particularly if no-one else exercises it. If we are able to use libmount, and there is a weird problem, I think it will be more straightforward for the user to debug because they can compare more directly with the behaviour of "/bin/mount".

https://twitter.com/mjg59/status/920496374549856256

That's just a suggestion though, I haven't inspected the kernel codepath for when only tmpfs is used, to see how trustworthy or not I think it is :).

yuwata commented:
@sourcejedi Thanks. Updated. Now we use libmount to parse mount flags from mount options.

yuwata commented:
@topimiettinen Still I am not clear your point. But, now the setting can take mount options. Thus, user can specify selinux context in the setting.

poettering commented on src/core/namespace.c:
```diff
@@ -131,6 +131,13 @@ static const MountEntry protect_home_read_only_table[] = {
+static const MountEntry protect_home_tmpfs_table[] = {
+        { "/home",               TMPFS,        true, .read_only = true, .option_const = "mode=0755" },
+        { "/run/user",           TMPFS,        true, .read_only = true, .option_const = "mode=0755" },
+        { "/root",               TMPFS,        true, .read_only = true, .option_const = "mode=0755" },
```
hmm, so if the user selects: `ProtectHome=tmpfs + ReadWriteDirectory=/root` then in your current implementation, /root would be mounted over, no?

I think the TMPFS entry in the MountMode structure should most likely move to the very end of the list, so that anything wins over it, and this will make such a /root bind mount work, among others

yuwata replied:
Indeed. Thanks. I will update this.

topimiettinen replied:
For `/root`, I'd use `mode=0700`.

yuwata replied:
Updated.

poettering commented on src/core/dbus-execute.c:
```diff
@@ -2244,6 +2245,47 @@ int bus_exec_context_set_transient_property(
+                        if (!path_is_absolute(i))
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid %s", name);
+
+                        path_kill_slashes(i);
```
hmm, it's bad enough we don't maintain these directory lists in fully parsed structures currently but I am very sure we should expose them fully normalized however these days, i.e. this property should be exposed as `a(bss)`, i.e. as array of triplets consisting of a boolean (for the ignore_enoent value), a path string and an option string...

The only reasons we don't expose ignore_enoent for the other options is that the boolean was added later on, and changing dbus property types after they have been introduced is nasty. But for new properties this doesn't apply, we really should expose them with full data, and fully normalized from day one.

yuwata replied:
OK. Updated.

sourcejedi commented:
I really like calling it TemporaryFileSystem= (and prominently mentioning the mode option), it makes it clearer how this will interact with SELinux, ACLs etc. which were set on the original directory.  IMO this mostly shifts the burden to SELinux people to tell us if this causes a horrible defeat of their system in some case, that they want us to leave a warning about :).

sourcejedi commented on src/core/namespace.c:
```diff
@@ -724,6 +796,27 @@ static int mount_empty_dir(const MountEntry *m) {
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get mount flags from mount options %s: %m", mount_entry_option(m));
+
+        if (mount("tmpfs", mount_entry_path(m), "tmpfs", flags, mount_entry_option(m)) < 0)
```
I'm confused :(.

Clearly your tests are passing, and you made sure to try and exercise the libmount function.  But you're still passing *all* the text options to the mount() system call, right?  When I asked if that was ok, it said no...

```
$ cat > test2.c
#include <sys/mount.h>

int main() {
        mount("tmp", "/mnt", "tmpfs", MS_NODEV|MS_STRICTATIME|MS_RDONLY, "ro");
        mount("tmp", "/mnt", "tmpfs", MS_NODEV|MS_STRICTATIME, "");
        mount("tmp", "/mnt", "tmpfs", MS_NODEV|MS_STRICTATIME|MS_RDONLY, "");
        mount("tmp", "/mnt", "tmpfs", MS_RDONLY, "ro");
}
$ make test2 >/dev/null
$ sudo strace -e trace=mount test2
mount("tmp", "/mnt", "tmpfs", MS_RDONLY|MS_NODEV|MS_STRICTATIME, "ro") = -1 EINVAL (Invalid argument)
mount("tmp", "/mnt", "tmpfs", MS_NODEV|MS_STRICTATIME, "") = 0
mount("tmp", "/mnt", "tmpfs", MS_RDONLY|MS_NODEV|MS_STRICTATIME, "") = 0
mount("tmp", "/mnt", "tmpfs", MS_RDONLY, "ro") = -1 EINVAL (Invalid argument)
```

Do you know why your tests are passing, and my paranoia was not valid?

sourcejedi replied:
In principle it would also be nice if the tests exercised one of the non-flag options as well.

yuwata replied:
Thank you for the example. In the updated version, all flags are removed from the option string.

topimiettinen commented:
@yuwata Specifying the mount options looks good to me, except the ACLs and extended & Linux attributes can't be specified. Maybe that's an acceptable limitation. I also like the new name.

Regarding #2780, does this permit nesting of directories, like
```
TemporaryFileSystem=/usr
ReadOnlyPaths=/usr/lib
```
?

topimiettinen commented:
What's actually the use case for `ProtectHome=tmpfs`? Some poorly designed service wants to access `/root`, `/run/user` or `/home`, but we don't care what it reads or writes as the contents will be discarded, and inaccessible directory is not OK? Do such services exist?

yuwata commented:
Updated. Now, all mount flags are removed from the option string. Also, DBus API is changed to `a(bss)`.

@topimiettinen
> TemporaryFileSystem=/usr
> ReadOnlyPaths=/usr/lib

I expect this works. The invoked processes can see nothing under `/usr` except `/usr/lib`. 

> What's actually the use case for ProtectHome=tmpfs?

E.g., when data for a web service is stored in /home/web, then it is natural that we want to hide user's home directories like /home/yuwata from httpd. Of course, we should not store such data under /home :-p

sourcejedi commented on src/core/namespace.c:
```diff
@@ -724,6 +752,20 @@ static int mount_empty_dir(const MountEntry *m) {
+static int mount_tmpfs(const MountEntry *m) {
+        assert(m);
+
+        /* First, get rid of everything that is below if there is anything. Then, overmount with our new empty dir */
```
Comment is a copy-paste from mount_empty_dir().  Should maybe be changed to "our new tmpfs".

yuwata replied:
Thanks. updated.

sourcejedi commented on src/core/namespace.c:
```diff
@@ -1305,6 +1359,79 @@ int bind_mount_add(BindMount **b, unsigned *n, const BindMount *item) {
+
+        a = flag_string;
+        while((a = strchr(a, ',')))
+                *a = '\0';
```
haha.  Yep, I think this is all sorted out nicely now.

poettering commented on src/core/namespace.c:
```diff
@@ -1305,6 +1359,79 @@ int bind_mount_add(BindMount **b, unsigned *n, const BindMount *item) {
+
+        a = flag_string;
+        while((a = strchr(a, ',')))
+                *a = '\0';
```
what is this loop supposed to do? i mean it looks for the first comma, patches it into an empty string and then looks in that empty string for more commands, which it's obviously not going to find... not sure i follow

yuwata replied:
Indeed! Thanks.

poettering commented on src/core/namespace.c:
```diff
@@ -1305,6 +1359,79 @@ int bind_mount_add(BindMount **b, unsigned *n, const BindMount *item) {
+
+        r = mnt_optstr_apply_flags(&flag_string, flags, mnt_get_builtin_optmap(MNT_LINUX_MAP));
+        if (r < 0)
+                return r;
```
I think we should move the mount option processing to some call in moun-util.[ch] or so, after all i figure nspawn should use that too in the long run... maybe some function: `mount_option_mangle(const char *options, unsigned long *ret_flag_options, char **ret_remaining_options)` that hides the mangling away, and removes all bit flag options from the string in one go

yuwata replied:
Yeah, nspawn-mount.c needs to be updated. I thought I will do in another PR. But, now, yes, I will do that in this PR.

yuwata replied:
Updated.

sourcejedi commented on test/test-execute/exec-temporaryfilesystem-ro.service:
```diff
@@ -0,0 +1,33 @@
+ExecStart=/bin/sh -c 'rm /tmp/thisisasimpletest-temporaryfilesystem'
+
+TemporaryFileSystem=/var:ro
+BindPaths=/tmp:/var/test-exec-temporaryfilesystem-rw
```
Sorry, another review dribble.  I noticed some subtle points by looking at this.

## Point 1

I think you pointed earlier that this is able to work correctly because we remount readonly in a second step.  I.e. only after we created /var/test-exec-temporaryfilesystem-rw, do we remount /var as readonly.

It's a pity the "add documents" commit would not let users realize that this works.  And this mode of operation is the ideal one for the original use case.  Maybe including an example of it in the documentation would be enough, at least to encourage users to try it and see that it works :).

## Not so subtle point 2

The original examples were like allowing access to /etc/bind but not allowing access to anything else in /etc.  This is quite subtle to do (I think there's similar comments elsewhere in the code), because if systemd first overmounts /etc, then it cannot easily access /etc/bind in order to bind-mount it.  If this works, we really really want it to be tested.  If this doesn't work, I think we don't want to close those bugs yet?

I had an idea earlier, to effectively enable the same code used for `RootDirectory=` even when the root directory is not different.  (Though it should be skipped when we're not doing _any_ namespace games).  `mount --rbind / /run/systemd/unit-root` is less expensive than you might think.  But it gets more complex, because currently we just use `mount --move /run/systemd/unit-root /`.  That means we leave _everything_ overmounted, it should work fine but it's ugly.  Ideally we would use `cd /run/systemd/unit-root && pivot_root . run/systemd/old-root`, and then `umount -l /run/systemd/old-root`.  (I originally tried umount -R /run/systemd/old-root` but [https://unix.stackexchange.com/questions/418100/umount-r-on-bind-mounts-takes-a-non-neglible-amount-of-time-why](found it caused an unexpected delay)).  I think my original motivation for this scheme was different and maybe sounds silly: if setup_namespace() fails, we currently ignore it and continue with the half-setup namespace.  My scheme would allow setup_namespace() to clean up with a simple `umount -l /run/systemd/unit-root`.

I had a WIP branch for this which consists of one code comment and two prep commits :).  https://github.com/sourcejedi/systemd/tree/rootdirectory

## Point 3

There's a mismatch between what the current remount-readonly code does, and what we would have written for this case specifically.  The current code is _bind_mounting the tmpfs read-only.  This is legal, it doesn't create an extra mount entry, and it we probably don't care about the difference.  As well as affecting the behaviour when it is bind mounted, the difference will be visible in /proc/self/mountinfo

    449 442 0:48 / /tmp ro,relatime - tmpfs tmpfs rw,seclabel

as opposed to

    448 443 0:48 / /mnt ro,relatime - tmpfs tmpfs ro,seclabel

sourcejedi replied:
My scheme in point 2 has the caveat that the `pivot_root()` syscall implementation is a bit funky.  The manpage for it points out that it involves scanning kernel threads to check if any of them need changing directory.  So I'm not 100% confident whether or not it's a good idea to call `pivot_root()` for anything other than switching from initramfs to the real system.

yuwata replied:
About point 3. With exec-temporaryfilesystem-ro.service, I get the following.
```
# /proc/self/mountinfo
438 400 0:58 / /var ro,nodev shared:226 - tmpfs tmpfs rw
# /proc/self/mounts
tmpfs /var tmpfs ro,nodev 0 0
```
It may be misleading. But, at least, this seems not in the scope of this PR, I think.

yuwata replied:
About point 2. Sorry, I have no idea about this... But your point seems about whole design of `setup_namespace()` rather than the new setting `TemporaryFileSystem=`. Right? If so, this is out of the scope this PR provides, I think.

sourcejedi replied:
> It may be misleading. But, at least, this seems not in the scope of this PR, I think, as this 'problem' affects not only TemporaryFileSystem= but also ReadOnlyPaths= or BindReadOnlyPaths=. Moreover, of course, I do not say that we do not need to fix this, but, unfortunately, I do not know how to fix this.

Thank you for looking.

1) I don't think we need to fix this.  I thought you might want to know though.  I'd still say it looks like the code _could_ have been written deliberately.

2) "Misleading" sounds a little strong.  That said, it doesn't 100% match the documentation.  I.e. when we set `ro` as a mount option, then "set as a mount option" is meaning something slightly different than for all the other mount options.

3) ReadOnlyPaths= and BindReadOnlyPaths= are both bind mounts, so it is expected that they work like `mount -o bind,ro olddir newdir`.  They cannot work any other way.

However TemporaryFileSystem= is a tmpfs mount.  So we have the opportunity of setting options on the underlying filesystem ("superblock"), instead of on the individual mount point.

In an ideal world, we would do the exact right thing, even on the obscure detail.  It would mean that if you play around, discover the obscure detail, and become confused - then your confusion will be because systemd did the right thing, not the wrong thing :).  And you will be happy to learn and have no annoyance at systemd :).

sourcejedi replied:
> About point 2. Sorry, I have no idea about this... But your point seems about whole design of setup_namespace() rather than the new setting TemporaryFileSystem=. Right? If so, this is out of the scope this PR provides, I think.

You need to have an idea about this, sorry :).  Please can you test the case by @topimiettinen 

    TemporaryFileSystem=/usr
    ReadOnlyPaths=/usr/lib

Can the service can see the full contents of /usr/lib?  If so, please make sure there is a test case that covers it.

If it doesn't work, you need to change the text of this PR to make sure that it does not close issues #7895, #7153, and #2780.

As you say, this could be addressed by a different PR.  (TemporaryFileSystem= could still be used for some purpose, but in a different way, matching how systemd uses it for the service private directories).  Sorry for making such a confusing comment about setup_namespace() :).  I am very pleased with my idea :) but the best explanation would be if I had working code to show you.

sourcejedi replied:
I am very sympathetic with the amount of work you have here, and not wanting further delays from out-of-scope work :).

topimiettinen replied:
For the httpd use case, should this work? 
```
ProtectHome=tmpfs
ReadOnlyPaths=/home/httpd
```

yuwata commented:
Updated. Adds `mount_option_mangle()`, and use it in namespace.c and `mount_verbose()` which is used by `nspawn` so many times.

topimiettinen commented on src/basic/mount-util.c:
```diff
@@ -874,3 +889,77 @@ int mount_propagation_flags_from_string(const char *name, unsigned long *ret) {
+         * See more examples in test-mount-utils.c.
+         *
+         * Note that *ret_mount_flags is not initialized in this function.
+         * The caller needs to initialized to zero or some default vaule. */
```
vaule -> value

poettering commented on src/basic/mount-util.c:
```diff
@@ -874,3 +878,77 @@ int mount_propagation_flags_from_string(const char *name, unsigned long *ret) {
+        assert(ret_mount_flags);
+        assert(ret_remaining_options);
+
+        mount_flags = *ret_mount_flags;
```
hmm, not a fan of parameters that are used for input and output like this... I think I'd prefer it if we'd pass this in in a separate input parameter, i.e.:

```c
int mount_option_mangle(const char *options, unsigned long mount_flags, char **ret_remaining_options, unsigned long *ret_mount_flags);
```

poettering commented on src/basic/mount-util.c:
```diff
@@ -874,3 +878,77 @@ int mount_propagation_flags_from_string(const char *name, unsigned long *ret) {
+                if (!ent->name) {
+                        char *t;
+
+                        t = strjoin(ret, isempty(ret) ? "" : ",", word);
```
looks like a perfect job for strextend()

poettering replied:
`strextend()` also has the nice benefit that you can drop the `strdup("")` call further up, as it handles NULL strings as first parameter just fine. That way we can save one allocation in the common case, and then do the strdup("") only when ret is still NULL at the end...

poettering commented:
ah, please always also add a brief comment when pushing a new version, github doesn't always notify on force pushes (no idea why...), so I didn't notice this early.

Patch looks great, I'd press the green button now, alas it needs a rebase first!

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -930,6 +930,19 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+        (<literal>:</literal>) and mount option such as <literal>size=10%</literal> or <literal>ro</literal>. If no
+        mount option is specified, then the tmpfs is mounted with <literal>mode=0755</literal>.
+        Note that each tmpfs enables <literal>nodev</literal> and <literal>strictatime</literal> options, that is,
+        each tmpfs is mounted with <literal>MS_NODEV|MS_STRICTATIME</literal> mount flags.
```
I think this sentence needs to say "by default" or similar.

AFAICT, from looking at the implementation commit and the test for mount_option_mangle(), it's possible to pass ":nodev,noatime" and override the default options.

sourcejedi commented:
Currently https://github.com/systemd/systemd/pull/7908#discussion_r168113252 is unanswered.  I think it will be important to have some confirmation of whether this fixes #7895, #7153, #2780.

This is my fault: I need to remember the Review button is required if I want to set a visible red/green marker (and not just as a way to batch comments on different lines).  Possibly I'm also not being decisive enough sometimes about choosing a red/green review instead of a "comment" one.

yuwata commented:
Sorry for late reply. I was busy last few days.

@poettering Now this is rebased. I did not comment last time, because it has not addressed comments by @sourcejedi and @topimiettinen. The current version still does not address them yet...

@sourcejedi and @topimiettinen Thank you for the comments. I will re-consider them and update this. Please wait a day, I hope. Sorry for late to answer them.

yuwata commented:
@sourcejedi I may completely misunderstand some of your points. I found this does not work with `Read{Write,Only}Paths=`, but works fine with `Bind{,ReadOnly}Paths=`. So, yeah, the issues are not fixed yet. Hmm, please give me more time to fix these issue...

In the updated version, tentatively all references to the issues in commit message are dropped, and the man page entries for the settings are updated.

sourcejedi commented:
```
Failed to mount /usr/lib64 to /run/systemd/unit-root/usr/lib64: No such file or directory
exec-temporaryfilesystem-usr.service: Failed to set up mount namespacing: No such file or directory
exec-temporaryfilesystem-usr.service: Failed at step NAMESPACE spawning /bin/sh: No such file or directory
Received SIGCHLD from PID 20678 ((sh)).
```

Nope.  /bin/sh is an alias for /usr/bin/sh.  Your /usr/bin is empty.  Because it's a bind mount of an empty directory /usr/bin on the tmpfs /usr.  Remember that we automatically create the directory for a mount point if it does not exist.

sourcejedi commented:
Hmm no, my hypothesis doesn't seem to be the specific reason for this error.  Interesting.

sourcejedi commented:
I see.  You have to have at least one bind mount in order to trip this condition

https://github.com/systemd/systemd/blob/65dd2d5b43ff36f363da77795cc11a3f260daa16/src/core/namespace.c#L1074

This is what enables the logic of setting up the namespace in `/run/systemd/unit-root` first, instead of applying directly to `/`.  As per my explanation, you really need this approach in order to fulfil this use case.

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -930,6 +930,37 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+        <varname>BindReadOnlyPaths=</varname>. See the example below.</para>
+
+        <para>Example: if a unit has the following,
+        <programlisting>TemporaryFileSystem=/var
```
I was trying to say that we should have an example which shows a tmpfs with `:ro`, and that it still works to have BindReadOnlyPaths= or whatever on paths inside it.  `:ro` is the more general configuration, for this use case, because it will work the same for `root`: it will stop you filling up RAM/swap space with files when you didn't realize the app was still configured to write to one of the tmpfs directories.

poettering commented:
> else if (root_image || n_bind_mounts > 0) {

This conditions should be updated I guess, to also trigger when there's at least one tmpfs, since unlikely the readonly/writable mounts the tmpfs mounts obstruct view on directories.

poettering commented on src/core/execute.c:
```diff
@@ -4181,6 +4189,28 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
+                        fprintf(f, "%sTemporaryFileSystem: %s%s:%s\n", prefix,
+                                t->ignore_enoent ? "-" : "",
+                                t->path,
+                                str);
```
hmm, for the dumping I'd probably just output the options string as it is, i.e. not process it with mnt_optstr_apply_flags or such

yuwata replied:
OK. Updated.

poettering commented on man/systemd.exec.xml:
```diff
@@ -930,6 +930,37 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+
+        <para>Note that when a path is specified with the mount option <literal>ro</literal>, then temporary file system is
+        once mounted there without <literal>ro</literal>, that is, with read-write mode, then it will remounted with
+        <literal>ro</literal> after setupping the directory tree under the path is finished. So, the corresponding entry in
```
"setupping" is not a word... "setting up" should be better

sourcejedi commented:
>     else if (root_image || n_bind_mounts > 0) {

> This conditions should be updated I guess, to also trigger when there's at least one tmpfs, since unlikely the readonly/writable mounts the tmpfs mounts obstruct view on directories.

-EPARSE, you meant "*unlike* the readonly/writeable mounts", not "unlikely".  But yes, I think that works fine.

poettering commented:
hmm, regarding the read-only thing: I think it would make sense to make both the superblock and the bind mount read-only, i.e. i'd ensure that both ro/rw appearances in /proc/self/mountinfo show up as "ro". Making one read-only should of course suffice, but it's easy to make both read-only hence I think we should, after all these are security features and two levels of security is nicer than one... or in other words we should both invoke MS_REMOUNT|MS_RDONLY|MS_BIND and MS_REMOUNT|MS_RDONLY on all tmpfs mount points after we set up in them whatever we needed to set up in them. The former will make the bind mount read-only and the latter the superblock.

In fact, we should do the same for PRIVATE_DEV entries (where we currently only make the superblock read-only, but not the bind mount) and the EMPTY_DIR entries (where we currently only make the bind mound read-only) and always make both bind mount + superblock readonly for them.

(in case you wonder about all of this: linux maintains one "superblock" in memory for each file system, and that can be read-only or not. A file system can then appear at multiple places in the file system hierarchy – each place is called a "bind mount". Each one of those can be made read-only or not too. Write access is only granted if the path you use to a file goes through a bind mount marked writable *and* points onto a superblock that is marked writable.  If either the bind mount or the superblock are read-only write access is denied. Now, when you invoke mount() and include MS_BIND in the mount flags you always refer to the bind mount's ro/rw flag, and if you do not you refer to the superblock's ro/rw flag. What I am proposing now is to always set both flags to read-only if we are the ones who create the superblock — which is the case for all fresh tmpfs instances we mount.)

poettering commented:
> This conditions should be updated I guess, to also trigger when there's at least one tmpfs, since unlikely the readonly/writable mounts the tmpfs mounts obstruct view on directories.
> 
> -EPARSE, you meant "unlike the readonly/writeable mounts", not "unlikely". But yes, I think that works fine.

Yes, of course, sorry for the confusion. The rule is that as soon as there's any mount included that might obstruct view on some paths, and if there are mounts that make arbitrary paths appear elsewhere we need to go the indirection route...

sourcejedi commented:
> Now, when you invoke mount() and include MS_BIND in the mount flags you always refer to the bind mount's ro/rw flag, and if you do not you refer to the superblock's ro/rw flag. What I am proposing now is to always set both flags to read-only if we are the ones who create the superblock — which is the case for all fresh tmpfs instances we mount.)

My conclusion when I tested, was that if you do not include MS_BIND, then the value of the MS_RDONLY flag is applied to *both* the superblock and the "bind mount".  So we only need to do one remount operation, which is nice.

```
$ unshare -rm
# strace -e trace=mount mount -ttmpfs -oro tmpfs /root
mount("tmpfs", "/root", "tmpfs", MS_MGC_VAL|MS_RDONLY, NULL) = 0
+++ exited with 0 +++
# grep /root /proc/self/mountinfo
260 84 0:50 / /root ro,relatime - tmpfs tmpfs ro,seclabel,uid=1001,gid=1001
```

yuwata commented:
@poettering and @sourcejedi Thank you for the comments. I've force-pushed an updated version. Please take a look. The main changes are below.
- Now tmpfs is remounted without MS_BIND at first, then remounted with MS_BIND later.
- Always set `mode=0755`. This can be overridden explicitly specify `mode=` option.
- Create a root directory even if no `BindPaths=` is specified.
- Options specified in a unit file are printed in `exec_context_dump()`.
- Update tests.
- Update man pages.

yuwata commented:
Updated.
- Drop '-' prefix for `TemporaryFileSystem=`.
- Make '-' prefix work for `Bind{ReadOnly}Paths=`

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -932,6 +932,29 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+
+        <para>Example: if a unit has the following,
+        <programlisting>TemporaryFileSystem=/var:ro
+BindReadOnlyPaths=/var/lib/systemd</programlisting>
```
I think we got `ReadOnlyPaths=` working, so we can use that in the example instead of the more verbose `BindReadOnlyPaths=`?

sourcejedi commented on src/core/namespace.c:
```diff
@@ -887,9 +944,15 @@ static int make_read_only(const MountEntry *m, char **blacklist, FILE *proc_self
+                        if (mount(NULL, mount_entry_path(m), NULL, MS_REMOUNT | MS_RDONLY | m->flags, mount_entry_options(m)) < 0)
+                                if (errno != ENOENT || !m->ignore)
+                                        return -errno;
+
```
(the call to bind_remount_recursive_with_mountinfo() below:) ah, hmm, doesn't this mean the `ro` tmpfs will also remount all the bind mounts which are underneath it to become read-only bind mounts?  I don't think we want that.

sourcejedi replied:
Ah, I see.  I missed the `blacklist` parameter that makes this work correctly.

sourcejedi replied:
I still think this will work fine without the second, MS_BIND remount of the tmpfs.  Having both calls shouldn't break anything though.

yuwata replied:
OK, now I agree with you. Calling `bind_remount_recursive_with_mountinfo()` for TMPFS and EMPTY_DIR is dropped.

poettering commented on src/core/dbus-execute.c:
```diff
@@ -2366,7 +2358,7 @@ int bus_exec_context_set_transient_property(
                                                 "%s=%s%s:%s:%s",
                                                 name,
-                                                ignore_enoent ? "-" : "",
+                                                !!ignore_enoent ? "-" : "",
```
not that it matters, but the `!!` here is redundant... `!!` is usually used to turn a non-boolean into a boolean value, i.e. 0 or 1. But the ternary operator does that anyway for the test...

poettering commented:
I am happy, I'd merge this as it is now. But I am not a native English speaker — @sourcejedi appears to be one however, hence I'll defer to him regarding the man page text. 

@sourcejedi anything left to do regarding the documentaiton? (or is there anything left on the code side?)

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -932,6 +932,29 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+      <varlistentry>
+        <term><varname>TemporaryFileSystem=</varname></term>
+
+        <listitem><para>Takes a space-separated list of mount points for temporary file system (tmpfs). If set, a new file
```
```
-Takes a space-separated list of mount points for temporary file system (tmpfs).
+Takes a space-separated list of mount points for temporary file systems (tmpfs).
```

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -788,14 +788,23 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+        <literal>tmpfs</literal>. If true, the directories <filename>/home</filename>, <filename>/root</filename> and
+        <filename>/run/user</filename> are made inaccessible and empty for processes invoked by this unit. If set to
+        <literal>read-only</literal>, the three directories are made read-only instead. If set to <literal>tmpfs</literal>,
+        temporary file system is mounted on the three directories with read-only mode. The value <literal>tmpfs</literal>
```
```
-temporary file system is mounted on the three directories with read-only mode
+temporary file systems are mounted on the three directories in read-only mode.
```

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -788,14 +788,23 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+        is useful to hide home directories not relevant to the processes invoked by the unit, while necessary directories
+        are still visible by combining with <varname>BindPaths=</varname> or <varname>BindReadOnlyPaths=</varname>.</para>
+
+        <para>Setting this <literal>yes</literal> is almost equivalent to set the three directories in
```
```
-Setting this <literal>yes</literal> is almost equivalent to set the three directories in
+Setting this to <literal>yes</literal> is mostly equivalent to setting the three directories in
```

sourcejedi commented on man/systemd.exec.xml:
```diff
@@ -788,14 +788,23 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+
+        <para>Setting this <literal>yes</literal> is almost equivalent to set the three directories in
+        <varname>InaccessiblePaths=</varname>. Similary, <literal>read-only</literal> is equivalent to
+        <varname>ReadOnlyPaths=</varname>, and <literal>tmpfs</literal> is to <varname>TemporaryFileSystem=</varname>.</para>
```
```
-is to <varname>TemporaryFileSystem=
+is mostly equivalent to <varname>TemporaryFileSystem=
```

yuwata commented:
@poettering and @sourcejedi Thank you for reviewing this so many times. I've force-pushed a new version. Main changes are
- drop `bind_remount_recursive_with_mountinfo()` for TMPFS and EMPTY_DIR,
- drop unnecessary `!!` to cast from int to bool, and
- man page fixes.

Note that still `Read{Write,Only}Paths=` cannot use with `TemporaryFileSystem=`. I have a plan to support this, but I want to do it in another PR.

sourcejedi commented:
The one CI instance is just another kernel GPF.

I'll be interested to see what's stopping `Read{Write,Only}Paths=`, I didn't guess it.  You can tag me in to that PR, if you're not fed up reading my comments :).  Ok, let's try merging this now.

evverx commented on test/test-execute/exec-temporaryfilesystem-options.service:
```diff
@@ -0,0 +1,11 @@
+
+# Check /proc/self/mountinfo
+ExecStart=/bin/sh -c 'test $$(awk \'$$5 == "/var" { print $$6 }\' /proc/self/mountinfo) = "ro,nodev,relatime"'
+ExecStart=/bin/sh -c 'test $$(awk \'$$5 == "/var" { print $$11 }\' /proc/self/mountinfo) = "ro,mode=700"'
```
I'm wondering why `test-execute` didn't fail on `Fedora`. There is usually `seclabel` in the 11th field of `/proc/self/mountinfo`, which makes it impossible for `ro,mode=700` to pass the check.

evverx replied:
```sh
exec-temporaryfilesystem-options.service: Passing 0 fds to service
exec-temporaryfilesystem-options.service: About to execute: /bin/sh -c 'test $$(awk '$$5 == "/var" { print $$6 }' /proc/self/mountinfo) = "ro,nodev,relatime"'
exec-temporaryfilesystem-options.service: Forked /bin/sh as 10524
Followed symlinks /run/systemd/unit-root/var → /run/systemd/unit-root/var.
Applying namespace mount on /run/systemd/unit-root/var
exec-temporaryfilesystem-options.service: Executing: /bin/sh -c 'test $(awk '$5 == "/var" { print $6 }' /proc/self/mountinfo) = "ro,nodev,relatime"'
Received SIGCHLD from PID 10524 (sh).
Child 10524 (sh) died (code=exited, status=0/SUCCESS)
exec-temporaryfilesystem-options.service: Child 10524 belongs to exec-temporaryfilesystem-options.service.
exec-temporaryfilesystem-options.service: Main process exited, code=exited, status=0/SUCCESS
exec-temporaryfilesystem-options.service: Running next main command for state start.
exec-temporaryfilesystem-options.service: Passing 0 fds to service
exec-temporaryfilesystem-options.service: About to execute: /bin/sh -c 'test $$(awk '$$5 == "/var" { print $$11 }' /proc/self/mountinfo) = "ro,mode=700"'
exec-temporaryfilesystem-options.service: Forked /bin/sh as 10527
Followed symlinks /run/systemd/unit-root/var → /run/systemd/unit-root/var.
Applying namespace mount on /run/systemd/unit-root/var
exec-temporaryfilesystem-options.service: Executing: /bin/sh -c 'test $(awk '$5 == "/var" { print $11 }' /proc/self/mountinfo) = "ro,mode=700"'
Received SIGCHLD from PID 10527 (sh).
Child 10527 (sh) died (code=exited, status=1/FAILURE)
exec-temporaryfilesystem-options.service: Child 10527 belongs to exec-temporaryfilesystem-options.service.
exec-temporaryfilesystem-options.service: Main process exited, code=exited, status=1/FAILURE
exec-temporaryfilesystem-options.service: Failed with result 'exit-code'.
exec-temporaryfilesystem-options.service: Changed start -> failed
exec-temporaryfilesystem-options.service: Unit entered failed state.
	PID: 10527
	Start Timestamp: Wed 2018-02-21 09:24:08 UTC
	Exit Timestamp: Wed 2018-02-21 09:24:08 UTC
	Exit Code: exited
	Exit Status: 1
Assertion 'service->main_exec_status.status == status_expected' failed at ../src/test/test-execute.c:74, function check(). Aborting.
Aborted
```

yuwata replied:
Ouch!

yuwata replied:
looks good. Can I include this in a PR that makes `ReadWritePaths=` can work with `TemporaryFileSystem=`.

sourcejedi replied:
You are welcome to take the above code, I just want to make sure the tests not left broken (on Fedora) for too long.  If you think the ReadWritePaths= will be quick to merge then I guess you can include them together.

yuwata replied:
I just changed my mind. This is now pushed as #8241. Please take a look.

evverx replied:
@sourcejedi, @yuwata thank you for looking at this. I've already fixed the test locally, but I hesitated to show it because, well, here it is:
```diff
diff --git a/test/test-execute/exec-temporaryfilesystem-options.service b/test/test-execute/exec-temporaryfilesystem-options.service
index 1d5d76c81..d63f66fd2 100644
--- a/test/test-execute/exec-temporaryfilesystem-options.service
+++ b/test/test-execute/exec-temporaryfilesystem-options.service
@@ -5,7 +5,13 @@ Description=Test for TemporaryFileSystem with mount options
 Type=oneshot

 # Check /proc/self/mountinfo
-ExecStart=/bin/sh -c 'test $$(awk \'$$5 == "/var" { print $$6 }\' /proc/self/mountinfo) = "ro,nodev,relatime"'
-ExecStart=/bin/sh -c 'test $$(awk \'$$5 == "/var" { print $$11 }\' /proc/self/mountinfo) = "ro,mode=700"'
+ExecStart=/usr/bin/perl -alne 'if ($$F[4] eq "/var") { \
+  my %%a = map { $$_ => 1 } split(",", $$F[10]); \
+  if (!@a{ro,"mode=700"}) {next}; \
+  my %%b = map { $$_ => 1 } split(",", $$F[5]); \
+  if (@b{ro,nodev,relatime}) {$$found=1; exit}; \
+} \
+END {exit($$found ? 0 : 1)}' \
+/proc/self/mountinfo

 TemporaryFileSystem=/var:ro,mode=0700,nostrictatime
```
Don't ask me how and why it works. I don't know :-)
